lib/ansible/modules/files/unarchive.py

```diff
@@ -856,12 +857,12 @@ def main():
         module.fail_json(msg="Source '%s' not readable, %s" % (src, to_native(e)))
 
     # is dest OK to receive tar file?
-    if not os.path.isdir(dest):
-        module.fail_json(msg="Destination '%s' is not a directory" % dest)
+    if not os.path.isdir(b_dest):
+        module.fail_json(msg="Destination '%s' is not a directory" % b_dest)
 
-    handler = pick_handler(src, dest, file_args, module)
+    handler = pick_handler(src, b_dest, file_args, module)
 
-    res_args = dict(handler=handler.__class__.__name__, dest=dest, src=src)
+    res_args = dict(handler=handler.__class__.__name__, dest=b_dest, src=src)
```

Comment:
You might save `dest` and `b_dest` as vars, as here you'll want to use `dest` as "native".  Anywhere it's going back out through `fail/exit_json` we want native.  This applies to the next several changes too.

Reply:
Note that fail/exit_json should be handling bytes or text; making them the proper string type for that python version before passing them in to the stdlib's json.dumps().